demos/speech_recognition_demo/python/ctcdecode-numpy/ctcdecode_numpy/ctc_beam_search_decoder.h

```diff
@@ -9,23 +9,80 @@
 #ifndef CTC_BEAM_SEARCH_DECODER_H_
 #define CTC_BEAM_SEARCH_DECODER_H_
 
+#include <memory>
 #include <string>
 #include <utility>
 #include <vector>
 
 #include "scorer_base.h"
 #include "output.h"
+#include "path_trie.h"
+
+
+class CtcDecoderState {
+public:
+  CtcDecoderState();
+  CtcDecoderState(const CtcDecoderState&) = delete;  // because copying PathTrie is not inplemented
+  CtcDecoderState(CtcDecoderState&&) = default;
+  CtcDecoderState& operator=(const CtcDecoderState& src) = delete;
+  CtcDecoderState& operator=(CtcDecoderState&& src) = default;
+
+  void init(
```

Comment:
This is actually close to mutable vs. unmutable objects. Two reasons: 1. I thought mutable-like objects are more convenient to organize thread pools with changing parameters (like LMs, alphabets, etc) -- they can be just held in a vector without any pointers. This can potentially be useful here. 2. In case a child class doesn't add new parameters, `init()` can be inherited.    ( And another general reason: Passing parameters to C++ constructor is inconvenient in some cases where user may be expected to want to conditionally initialize class inside `if/else` without pointers. )